client/host_queue.go

```diff
@@ -183,64 +183,56 @@ func (q *queue) rotateOpsWithLock() []op {
 			case *writeOperation:
 				namespace := v.namespace
-				namespaceKey := namespace.Hash()
-				currWriteOps = currWriteOpsByNamespace[namespaceKey]
-				currBatchElements = currBatchElementsByNamespace[namespaceKey]
-				if currWriteOps == nil {
-					currWriteOps = q.opsArrayPool.Get()
-					currBatchElements = q.writeBatchRawRequestElementArrayPool.Get()
+				idx := currWriteOpsByNamespace.indexOf(namespace)
+				if idx == -1 {
+					value := namespaceWriteBatchOps{
+						namespace:                            namespace,
+						opsArrayPool:                         q.opsArrayPool,
+						writeBatchRawRequestElementArrayPool: q.writeBatchRawRequestElementArrayPool,
+					}
+					idx = len(currWriteOpsByNamespace)
+					currWriteOpsByNamespace = append(currWriteOpsByNamespace, value)
 				}
 
-				currWriteOps = append(currWriteOps, ops[i])
-				currBatchElements = append(currBatchElements, &v.request)
-				currWriteOpsByNamespace[namespaceKey] = currWriteOps
-				currBatchElementsByNamespace[namespaceKey] = currBatchElements
+				currWriteOpsByNamespace.appendAt(idx, ops[i], &v.request)
 
-				if len(currWriteOps) == writeBatchSize {
+				if currWriteOpsByNamespace.lenAt(idx) == writeBatchSize {
 					// Reached write batch limit, write async and reset
-					q.asyncWrite(namespace, currWriteOps, currBatchElements)
-					currWriteOpsByNamespace[namespaceKey] = nil
-					currBatchElementsByNamespace[namespaceKey] = nil
+					q.asyncWrite(namespace, currWriteOpsByNamespace[idx].ops,
+						currWriteOpsByNamespace[idx].elems)
+					currWriteOpsByNamespace.resetAt(idx)
 				}
 			case *writeTaggedOperation:
 				namespace := v.namespace
-				namespaceKey := namespace.Hash()
-				currTaggedWriteOps = currTaggedWriteOpsByNamespace[namespaceKey]
-				currTaggedBatchElements = currTaggedBatchElementsByNamespace[namespaceKey]
-				if currTaggedWriteOps == nil {
-					currTaggedWriteOps = q.opsArrayPool.Get()
-					currTaggedBatchElements = q.writeTaggedBatchRawRequestElementArrayPool.Get()
+				idx := currTaggedWriteOpsByNamespace.indexOf(namespace)
+				if idx == -1 {
+					value := namespaceWriteTaggedBatchOps{
+						namespace:                                  namespace,
+						opsArrayPool:                               q.opsArrayPool,
+						writeTaggedBatchRawRequestElementArrayPool: q.writeTaggedBatchRawRequestElementArrayPool,
+					}
+					idx = len(currTaggedWriteOpsByNamespace)
+					currTaggedWriteOpsByNamespace = append(currTaggedWriteOpsByNamespace, value)
 				}
 
-				currTaggedWriteOps = append(currTaggedWriteOps, ops[i])
-				currTaggedBatchElements = append(currTaggedBatchElements, &v.request)
-				currTaggedWriteOpsByNamespace[namespaceKey] = currTaggedWriteOps
-				currTaggedBatchElementsByNamespace[namespaceKey] = currTaggedBatchElements
+				currTaggedWriteOpsByNamespace.appendAt(idx, ops[i], &v.request)
 
-				if len(currTaggedWriteOps) == writeBatchSize {
+				if currTaggedWriteOpsByNamespace.lenAt(idx) == writeBatchSize {
```

Comment:
noticed a bug unrelated to your changes: could you use `writeTaggedBatchSize` here

Reply:
Ah good call, sure thing I'll update this.

Reply:
Hm, actually there's no such option `WriteTaggedBatchSize` - is there any reason why we wouldn't just have a single value that applies to both?  We don't want our config to be overly complex whenever possible.

Reply:
Yeah I can see the value in having them the same. Opened an issue to consolidate any others that need to be: https://github.com/m3db/m3db/issues/534